gpbft/gpbft.go

```diff
@@ -1090,22 +1122,29 @@ type ConvergeTicket struct {
 
 func newConvergeState() *convergeState {
 	return &convergeState{
+		senders: map[ActorID]struct{}{},
 		values:  map[ChainKey]ConvergeValue{},
 		tickets: map[ChainKey][]ConvergeTicket{},
 	}
 }
 
 // Receives a new CONVERGE value from a sender.
+// Ignores any subsequent value from a sender from which a value has already been received.
 func (c *convergeState) Receive(sender ActorID, value ECChain, ticket Ticket, justification *Justification) error {
+	if _, ok := c.senders[sender]; ok {
+		return nil
+	}
+	c.senders[sender] = struct{}{}
 	if value.IsZero() {
 		return fmt.Errorf("bottom cannot be justified for CONVERGE")
 	}
 	key := value.Key()
 
-	// !!! TODO: Drop duplicates (see https://github.com/filecoin-project/go-f3/issues/12)
-	c.values[key] = ConvergeValue{Chain: value, Justification: justification}
-	c.tickets[key] = append(c.tickets[key], ConvergeTicket{Sender: sender, Ticket: ticket})
-
+	// Keep only the first justification and ticket received for a value.
+	if _, found := c.values[key]; !found {
+		c.values[key] = ConvergeValue{Chain: value, Justification: justification}
+		c.tickets[key] = append(c.tickets[key], ConvergeTicket{Sender: sender, Ticket: ticket})
+	}
```

Comment:
I think I missed this part of the conversation. The code looks correct, but I don't have a strong enough understanding of the protocol to give it an OK from that perspective.

Reply:
This is ok!